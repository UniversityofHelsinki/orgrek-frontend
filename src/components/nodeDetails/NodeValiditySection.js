import React from 'react';
import Validity from '../attributes/Validity';
import EditableAccordion from '../EditableAccordion';
import { useTranslation } from 'react-i18next';
import Placeholder from '../Placeholder';
import EditableContent from '../EditableContent';
import { authActions } from '../../auth';
import NodeValidityEditor from './NodeValidityEditor';
import { nodeValiditySchema } from '../../utils/validations';
import {
  useGetNodeValidityQuery,
  useSaveNodeValidityMutation,
} from '../../store';
import { useNodeId } from '../../hooks/useNodeId';

const NodeValiditySection = () => {
  const { t } = useTranslation();
  const nodeId = useNodeId();
  const { data: node, isFetching } = useGetNodeValidityQuery(nodeId);
  const saveNodeValidity = useSaveNodeValidityMutation();

  const empty = !node.startDate && !node.endDate;
  const title = t('valid_dates');

  const handleSubmit = (values) => {
    return saveNodeValidity({ data: values, id: nodeId }).unwrap();
  };

  return (
    <EditableAccordion
      title={title}
      loading={isFetching}
      defaultExpanded={!empty}
    >
      <EditableContent
        editorComponent={<NodeValidityEditor />}
        validationSchema={nodeValiditySchema}
        initialValues={node}
        onSubmit={handleSubmit}
        successMessage={t('nodeValidity.saveSuccess')}
        errorMessage={t('nodeValidity.saveError')}
        authActions={authActions.nodeValidity}
      >
        <Placeholder empty={empty} placeholder={<Validity />}>
          <Validity startDate={node.startDate} endDate={node.endDate} />
        </Placeholder>
      </EditableContent>
    </EditableAccordion>
  );
};

export default NodeValiditySection;
