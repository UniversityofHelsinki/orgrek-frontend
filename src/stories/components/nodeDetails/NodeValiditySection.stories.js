import {
  mockGetNodeValidity,
  mockSaveNodeValidity,
  withMockStore,
} from '../../../mockStore';
import NodeValiditySection from '../../../components/nodeDetails/NodeValiditySection';
import { userEvent, waitFor, within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

const now = new Date('2023-03-22T14:28:00+0200');
const nodeId = '1';

export default {
  component: NodeValiditySection,
  parameters: {
    systemTime: now,
    reactRouter: {
      searchParams: {
        uid: nodeId,
      },
    },
  },
};

export const Default = {
  decorators: [withMockStore()],
  parameters: {
    msw: {
      handlers: [
        mockGetNodeValidity(nodeId, {
          startDate: '2000-01-01',
          endDate: null,
        }),
        mockSaveNodeValidity(nodeId),
      ],
    },
  },
};
export const Empty = {
  ...Default,
  parameters: {
    msw: {
      handlers: [
        mockGetNodeValidity(nodeId, {
          startDate: null,
          endDate: null,
        }),
        mockSaveNodeValidity(nodeId),
      ],
    },
  },
};

export const EditMode = {
  ...Default,
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Muokkaa')).toBeInTheDocument();
    });

    await userEvent.click(canvas.getByText('Muokkaa'));
  },
};
