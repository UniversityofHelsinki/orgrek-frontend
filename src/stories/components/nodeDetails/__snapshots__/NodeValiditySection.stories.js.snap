// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/nodeDetails/NodeValiditySection Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.1.2000 alkaen",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-nodevaliditysection--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/NodeValiditySection Empty smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-nodevaliditysection--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
